ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/WebSocketTest.kt

```diff
@@ -74,6 +80,71 @@ class WebSocketTest : ClientLoader() {
         }
     }
 
+    @Test
+    fun testEchoWSS() = clientTests(ENGINES_WITHOUT_WEBSOCKETS + "Js" + "native:CIO") {
+        config {
+            install(WebSockets)
+        }
+
+        test { client ->
+            client.webSocket("wss://echo.websocket.org") {
+                outgoing.send(Frame.Text("PING"))
+                val frame = incoming.receive()
+                assertTrue(frame is Frame.Text)
+                assertEquals("PING", frame.readText())
+            }
+        }
+    }
+
+    @Test
+    fun testWsHandshakeHeaders() =
+        clientTests(ENGINES_WITHOUT_WEBSOCKETS + "native:CIO") {
+            if (PlatformUtils.IS_BROWSER) return@clientTests // browser websocket client does not support custom headers so the test gets ignored
+            config {
+                install(WebSockets)
+            }
+
+            test { client ->
+                val session = client.webSocketSession {
+                    url("$TEST_WEBSOCKET_SERVER/websockets/headers")
+                    header(CUSTOM_HEADER, CUSTOM_HEADER_VALUE)
+                }
+
+                val frame = session.incoming.receive()
+                assertTrue(frame is Frame.Text)
+                val headers =
+                    Json.decodeFromString<Map<String, List<String>>>(frame.readText())
+                val header = headers[CUSTOM_HEADER]?.first()
+                assertEquals(CUSTOM_HEADER_VALUE, header)
+            }
+        }
+    }
+
+    @Test
+    fun testWsHandshakeHeadersWithMultipleValues() =
+        clientTests(ENGINES_WITHOUT_WEBSOCKETS + "native:CIO") {
+            if (PlatformUtils.IS_BROWSER) return@clientTests // browser websocket client does not support custom headers so the test gets ignored
+            config {
+                install(WebSockets)
+            }
+
+            test { client ->
+                val session = client.webSocketSession {
+                    url("$TEST_WEBSOCKET_SERVER/websockets/headers")
+                    header(CUSTOM_HEADER, CUSTOM_HEADER_VALUE)
+                    header(CUSTOM_HEADER, CUSTOM_HEADER_VALUE)
+                }
+
+                val frame = session.incoming.receive()
+                assertTrue(frame is Frame.Text)
+                val headers =
+                    Json.decodeFromString<Map<String, List<String>>>(frame.readText())
+                val header = headers[CUSTOM_HEADER]
+                assertEquals(listOf(CUSTOM_HEADER_VALUE, CUSTOM_HEADER_VALUE), header)
```

Comment:
Also your test `testWsHandshakeHeadersWithMultipleValues` failed with `AssertionError: Expected <[custom_header_value, custom_header_value]>, actual <[custom_header_value,custom_header_value]>.` because you join all same headers using `joinToString(","), so the same header would not return `listOf(headerValue1, headerValue2)` but instead `listOf("$headerValue1,$headerValue2")`.
The list in your test has a size of 1 with one String containing both headers. Fix your test or fix the response in the Testserver, eg by using flattenEntries(), each pair has a unique String key and a String value containing all headerValues or use `toMap.mapValues { it -> it.first() }` which results into Map<String, String>.

Reply:
Sorry for delaying it so much. I have no idea why but IDEA doesn't seem to like that project. now it's not even recognized as a Gradle Project anymore

Reply:
I pushed a possible fix but since importing is stuck at
> warning workspace-aggregator-6d4bb098-5ecb-4134-acb8-43e7347bc4a3 > ktor-ktor-http-jsLegacy-test > webpack > micromatch > snapdragon > source-map-resolve > resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
warning workspace-aggregator-6d4bb098-5ecb-4134-acb8-43e7347bc4a3 > ktor-ktor-http-jsLegacy-test > webpack > micromatch > snapdragon > source-map-resolve > urix@0.1.0: Please see https://github.com/lydell/urix#deprecated

i can't test it